apps/src/templates/progress/progressHelpers.js

```diff
@@ -180,6 +241,176 @@ export const processedLevel = level => {
     levelNumber: level.kind === LevelKind.unplugged ? undefined : level.title,
     isConceptLevel: level.is_concept_level,
     bonus: level.bonus,
-    sublevels: level.sublevels
+    sublevels:
+      level.sublevels &&
+      level.sublevels.map(level => processedLevel(level, true))
+  };
+};
+
+/**
+ * Given a level progress object that we get from the server using either
+ * /api/user_progress or /dashboardapi/section_level_progress,
+ * extracts the result, appropriately discerning a locked/submitted
+ * result for certain levels.
+ */
+const getLevelResult = serverProgress => {
+  if (serverProgress.status === LevelStatus.locked) {
+    return TestResults.LOCKED_RESULT;
+  }
+  if (serverProgress.readonly_answers) {
+    return TestResults.READONLY_SUBMISSION_RESULT;
+  }
+  if (serverProgress.submitted) {
+    return TestResults.SUBMITTED_RESULT;
+  }
+
+  return serverProgress.result || TestResults.NO_TESTS_RUN;
+};
+
+export const levelProgressFromServer = serverObject => {
+  return {
+    status: serverObject.status || LevelStatus.not_tried,
+    result: getLevelResult(serverObject),
+    paired: serverObject.paired || false,
+    timeSpent: serverObject.time_spent || 0
   };
 };
+
+export const processServerStudentProgress = serverStudentProgress => {
+  return _.mapValues(serverStudentProgress, progress =>
+    levelProgressFromServer(progress)
+  );
+};
+
+export const processServerSectionProgress = serverSectionProgress => {
+  const studentProgress = _.mapValues(serverSectionProgress, student =>
+    processServerStudentProgress(student)
+  );
+  return studentProgress;
+};
+
+export const levelProgressWithStatus = status => {
+  return levelProgressFromServer({status: status});
+};
+
+export const levelProgressFromResult = result => {
+  return levelProgressWithStatus(activityCssClass(result));
+};
+
+export const mergeLevelProgressWithResult = (progress, result) => {
+  if (!progress) {
+    return levelProgressFromResult(result);
+  }
+  const mergedResult = mergeActivityResult(progress.result, result);
+  if (mergedResult === result) {
+    return progress;
+  }
+  return levelProgressFromResult(mergedResult);
+};
+
+const lessonFromServer = (serverObject, stageNumber, overrides = {}) => {
+  return {
+    name: overrides.name || serverObject.name,
+    id: overrides.id || serverObject.id,
+    lockable: overrides.lockable || serverObject.lockable,
+    lesson_plan_html_url:
+      overrides.lesson_plan_html_url || serverObject.lesson_plan_html_url,
+    description_student:
+      overrides.description_student || serverObject.description_student,
+    description_teacher:
+      overrides.description_teacher || serverObject.description_teacher,
```

Comment:
(NIT - mostly just curious) Here, we keep the underscores, but in other places (i.e. `levelGroupFromServer`) we change to camelcase. Is that intentional? Any particular reason?

Reply:
only reason here was to not refactor more than i had to. we used to be parsing the lesson data using `pick` (see the removed `progressRedux.lessonFromStage` at old line 499), which preserved the underscores. i needed to move away from that in order to support overrides here, but didn't deem it worthwhile to refactor all uses of those property names. want me to?

Reply:
No need. Thanks for the explanation!